samza-core/src/main/java/org/apache/samza/container/RunLoop.java

```diff
@@ -239,8 +258,29 @@ private void runTasks(IncomingMessageEnvelope envelope) {
     if (!shutdownNow) {
       if (envelope != null) {
         PendingEnvelope pendingEnvelope = new PendingEnvelope(envelope);
-        for (AsyncTaskWorker worker : sspToTaskWorkerMapping.get(envelope.getSystemStreamPartition())) {
-          worker.state.insertEnvelope(pendingEnvelope);
+        SystemStreamPartition sspOfEnvelope = null;
+        // when elasticity is enabled
+        // the tasks actually consume a keyBucket of the ssp.
+        // hence use the SSP with keybucket to find the worker(s) for the envelope
+        if (elasticityFactor <= 1) {
+          sspOfEnvelope = envelope.getSystemStreamPartition();
+        } else {
+          sspOfEnvelope = envelope.getSystemStreamPartition(elasticityFactor);
+          log.trace("elasticity enabled. using the ssp of the envelope as {}", sspOfEnvelope);
+        }
```

Comment:
Cant this be one call, so that 
envelope.getSSP(elasticity) returns the current enevelope.getSSP() with elasticity = 1, 
that way method, 
envelope.getSSP() can be deleted?

Reply:
done. 
But can not delete the envelope.getSSP() method as it is used by many samza components which have no idea or access to elasticityFactor.